api/src/main/java/org/pmiops/workbench/reporting/ReportingUploadServiceImpl.java

```diff
@@ -87,24 +87,53 @@ public void uploadSnapshot(ReportingSnapshot reportingSnapshot) {
     final List<InsertAllRequest> insertAllRequests = getInsertAllRequests(reportingSnapshot);
     final StringBuilder performanceStringBuilder = new StringBuilder();
     for (InsertAllRequest request : insertAllRequests) {
-      stopwatch.start();
-      final InsertAllResponse currentResponse = bigQueryService.insertAll(request);
-      responseMapBuilder.put(request.getTable(), currentResponse);
-      stopwatch.stop();
-      performanceStringBuilder
-          .append(
-              LogFormatters.rate(
-                  String.format("Streaming upload into %s table", request.getTable().getTable()),
-                  stopwatch.elapsed(),
-                  request.getRows().size(),
-                  "rows"))
-          .append("\n");
-      stopwatch.reset();
+      issueInsertAllRequest(stopwatch, responseMapBuilder, performanceStringBuilder, request);
     }
     log.info(performanceStringBuilder.toString());
     checkResponseAndRowCounts(reportingSnapshot, responseMapBuilder.build());
   }
 
+  /**
+   * Batch uploads {@link ReportingWorkspace}.
+   *
+   * <p>TODO(yonghao): checkResponseAndRowCounts for batch.
+   */
+  @Override
+  public void uploadBatchWorkspace(List<ReportingWorkspace> batch, long captureTimestamp) {
+    final Stopwatch stopwatch = stopwatchProvider.get();
+    final ImmutableMultimap.Builder<TableId, InsertAllResponse> responseMapBuilder =
+        ImmutableMultimap.builder();
+    final StringBuilder performanceStringBuilder = new StringBuilder();
+    final InsertAllRequest insertAllRequest =
+        workspaceRequestBuilder.build(
+            getTableId(WorkspaceColumnValueExtractor.class),
+            batch,
+            getFixedValues(captureTimestamp));
+    issueInsertAllRequest(
+        stopwatch, responseMapBuilder, performanceStringBuilder, insertAllRequest);
+    log.info(performanceStringBuilder.toString());
+  }
+
+  private void issueInsertAllRequest(
```

Comment:
javadoc here would be helpful. This function is at the crux of the whole shebang.